Fixes traits() with ES6 classes (STILL BROKEN)

fatcerberus opened the pull request:
ES6 classes require `new` which means the usual ES5 idiom of constructor forwarding using `base_ctor.apply(this)` doesn't work.  The only way to make it work is to use `Reflect.construct()` and shallow-copy the own properties over manually.

**Important:** This is currently still broken; it fails a supertrait test with`"#<C> is not a constructor`.  I haven't figured out what causes that yet.

matthewadams commented:
Thanks.  I'll try to have a look as soon as I have time.  Hopefully, we can knock this out.

fatcerberus commented:
I suspected at first that the issue might be `abstract class` not being newable but it turns out tsc simply transpiles that to a normal class, so I don’t think that’s it.

matthewadams commented:
Yeah, no such thing as `abstract` in JS...

fatcerberus commented on src/main/trait.ts:
```diff
@@ -12,7 +12,13 @@ export default function traits<T extends Array<Class<any>>> (constructors: T): C
       constructors.forEach((c: any) => {
-        Reflect.construct(c, [], this)
+        const tmp = Reflect.construct(c, [], this)
+        for (let key in Object.keys(tmp)) {
```
Oops, this for should be `for..of`, not `in`. I’ll fix it later, once I’m at home again.

fatcerberus commented:
I was discussing this with a friend offline - the third argument to `Reflect.construct()` should have been `new.target`, not `this`.  Fixing that causes a different test failure though; I’ll need to investigate further.

matthewadams commented:
I have to go offline for the night, sorry.

fatcerberus committed:
Fixes traits() with ES6 classes (STILL BROKEN)

ES6 classes require `new` which means the usual ES5 idiom of constructor
forwarding using `base_ctor.apply(this)` doesn't work.  The only way to
make it work is to use `Reflect.construct()` and shallow-copy the own
properties over manually.

fatcerberus commented:
Good thing Git collaboration doesn't require everyone to be online at all times, right? 😉 

`Reflect.construct()` indeed works here; the third argument should be `new.target`.  Another issue was that copying methods didn't work because it used Object.assign to do so; class methods are not enumerable.  I fixed it by using the same method of copying property descriptors as used for own properties.

These latest changes (see commit above for full diff) gets us about 90% of the way there; the final sticking point is that subclassing currently doesn't work.  The test fails with `c.bleep is not a function` because only a shallow copy of the trait classes has been done--so it finds the methods from `Subtrait` and `AnotherTrait` but not either of the `Supertrait`s.

Trouble is, `cls.__proto__` (and `cls.prototype.__proto__`) can only point to one class, meaning the only option left is to recursively copy properties and methods off all the classes in the prototype chain of each constructor.  That should let you pass the method call tests; however, it will *not* make `instanceof` work the way you want because the instanceof algorithm relies on being able to find the specified class in the prototype chain.

This is definitely an interesting challenge.

matthewadams commented:
I think the subclass factory idea should be our saving grace here, since it can be done in JavaScript, and the only consideration is that we just need a typesafe way of expressing it in TypeScript.  The key concept in `mutrait` is that a trait is a function that takes a class or `undefined` and returns a new class that is a subclass of the given class or `class {}`, respectively, that has the desired trait state & behavior.  I'm a TypeScript noob, but not a programming noob by any means.  Is there a simpler way of saying that in TypeScript?  The current `src/main/trait.ts` is most certainly _not_ for the faint of heart.

Thinking out loud here, given `type Class<T> = Function & { prototype: T }`, does it help if we say `type Trait = <T, S>(superclass?: Class<S>) => Class<T & S>`?  Would an instance of `Class<T & S>` be `instanceof` both `T` & `S`?

I'm not sure this'll help, but my intuition is telling me that it will...

fatcerberus commented:
It helps to not think of TypeScript as a separate language, I think - all you’re doing is adding type info to a codebase that’s fundamentally still JS which can then be verified by the compiler and promptly thrown away.  They don’t emit different code for different types (“no type-directed emit”) and even something like a cast has zero runtime effect.  TS is just kind of like a linter on steroids. :stuck_out_tongue:

If you can do it in JS, you can do it in TS the exact same way.  The question just becomes if you can represent what you want to do in a way the type system understands.  Which you always can, because the `any` type exists (but it’s preferred not to take the easy way out if you can avoid it, of course).

But that brings me to the issue here: `instanceof`.  JavaScript, even ES6 with its classes, is fundamentally a single-inheritance design.  An object only has one `.__proto__` which points to an object which itself only has one `.__proto__`, etc.  There’s no way to represent an object which inherits from multiple different things.

The trick Mutrait seems to do is to *override* `Symbol.hasInstance` — a trick that I should point out *only works in ES6+*.  You’ve said you’re fine with requiring ES6, correct?

matthewadams commented:
Yes, ES6+ is fine, since this is backend, server-side code.

The only reason `mutrait` uses `Symbol.hasInstance` is so that you can ask if a class has a prototype in its chain that got there by applying a trait _function_ to it, otherwise that question doesn't really make sense.

It'd be like asking this:  given `class Foo {}` & `function Bar() {}`, `new Foo() instanceof Bar`, which would _always_ return `false`.  However, if function `Bar` were a _trait_ (because traits are functions that return classes), then, with some extra bits that override `Symbol.hasInstance`, you _can_ ask if `new Foo() instanceof Bar`.

Doing this makes it natural for someone to determine if the instance you have in your hand effectively "implements" the interface represented by the trait.

matthewadams commented:
I was trying to take the following & run with it.
```
type Class<T> = Function & { prototype: T }
type Trait<T,SUP = {}> = (superclass?: Class<SUP>) => Class<T & SUP>
```

fatcerberus commented:
Yes, I was aware - I was under the impression you wanted `instanceof` to work for traits *without* overriding  Symbol.hasInstance - which turns out to be actually impossible for any class which `extends` more than one trait.

But this also means it's impossible to implement multiple traits without *some* copying going on.  mutrait claims it does everything without any copying, and I indeed didn't see any copying happening in the codebase when I skimmed through it, so I'm quite confused how that works.

w.r.t. this:
```ts
type Class<T> = Function & { prototype: T }
type Trait<T,SUP = {}> = (superclass?: Class<SUP>) => Class<T & SUP>
```
Your code doesn't actually care about this.  Only people reading the code and the TypeScript compiler do.  So even if you say something is `Class<T & S>`, that doesn't automatically make it `instanceof` T and S--because there's no way to represent that in the JS VM (unless T is a subclass of S or vice versa).

matthewadams commented:
Have a look at https://github.com/matthewadams/typescript-trait/pull/2

I'm trying to take a different approach, based a little bit on `mutrait`& a little bit on AspectJ, where each trait is represented by _both_ an interface and a trait function that returns a class that extends a given superclass and also implements the trait's interface.

I think I'm pretty close to getting it to work.
